### ISSUE TITLE
MIMEdefang filter process crashes when name of attachment being dropped is in UTF-8

### ISSUE THREAD

**Issue by jikamens** [The-McGrail-Foundation/MIMEDefang#54]
If someone emails you a ZIP file whose name as specified in the MIME headers in UTF-8, then MIMEdefang crashes when trying to drop the ZIP file and add a warning about it to the message. Example logs:

```
Feb 10 02:05:27 jik4 mimedefang.pl[4134278]: 21A75PbG055031: MDLOG,21A75PbG055031,bad_filename,nova narud%17Eba pdf.zip,application/zip,<info@aas.ba>,<elided>,=?UTF-8?Q?Re=3A_nova_narud=C5=BEba?=
Feb 10 02:05:27 jik4 mimedefang.pl[4134278]: 21A75PbG055031: MDLOG,21A75PbG055031,mail_in,,,<info@aas.ba>,<elided>,=?UTF-8?Q?Re=3A_nova_narud=C5=BEba?=
Feb 10 02:05:27 jik4 mimedefang-multiplexor[3749123]: 21A75PbG055031: Worker 0 stderr: open body: Invalid argument at /usr/share/perl5/vendor_perl/MIM         
Feb 10 02:05:27 jik4 mimedefang-multiplexor[3749123]: 21A75PbG055031: Worker 0 stderr: E/Entity.pm line 1892.                                                  
Feb 10 02:05:27 jik4 mimedefang-multiplexor[3749123]: Worker 0 died prematurely -- check your filter rules                                                     
Feb 10 02:05:27 jik4 mimedefang-multiplexor[3749123]: Reap: worker 0 (pid 4134278) exited normally with status 22 (WORKER DIED UNEXPECTEDLY)                   
Feb 10 02:05:27 jik4 mimedefang-multiplexor[3749123]: Worker 0 resource usage: req=304, scans=295, user=3.179, sys=0.240, nswap=0, majflt=600, minflt=13458, maxrss=48816, bi=6040, bo=46080                                                  
Feb 10 02:05:27 jik4 mimedefang[3749140]: 21A75PbG055031: Error from multiplexor: ERR No response from worker                                                  
```

The relevant MIME headers from the attachment in the spam message logged above:

```
Content-Transfer-Encoding: base64                                               
Content-Type: application/zip;                                                  
 name="=?UTF-8?Q?nova_narud=C5=BEba_pdf=2Ezip?="                                
Content-Disposition: attachment;                                                
 filename*=UTF-8''nova%20narud%C5%BEba%20pdf.zip;                               
 size=705397                                                                    
```

I don't know if this is the _right_ fix—in fact, I'm pretty sure that it's not—but it fixed the issue for me, so perhaps it will give someone else who knows more about all this stuff than I do an idea what the right fix should be:

```
--- mimedefang.pl	2022/02/10 16:46:42	1.1
+++ mimedefang.pl	2022/02/10 16:53:29
@@ -75,6 +75,7 @@
 use Sys::Hostname;
 use File::Spec qw ();
 use Errno qw(ENOENT EACCES);
+use Encode;
 
 # Detect these Perl modules at run-time.  Can explicitly prevent
 # loading of these modules by setting $Features{"xxx"} = 0;
@@ -989,10 +990,17 @@
 
     my ($part);
 
+    my $charset = undef;
+    if (Encode::is_utf8($data)) {
+        $data = encode 'UTF-8', $data;
+        $charset = 'utf-8';
+    }
+
     $part = MIME::Entity->build(Type => $type,
 				Top => 0,
 				'X-Mailer' => undef,
 				Encoding => $encoding,
+                                Charset => $charset,
 				Data => ["$data"]);
     defined ($fname) && $part->head->mime_attr("Content-Type.name" => $fname);
     defined ($disposition) && $part->head->mime_attr("Content-Disposition" => $disposition);
```

**Comment by bigio** [comment]
Which MIMEDefang version are you using ?

**Comment by jikamens** [comment]
2.85 From CentOS Stream 8

**Comment by bigio** [comment]
This has been fixed in 2.86 in a different way with commit 456aeb107d5bb8fd56cc440cb10adccc65bb55d2
Thanks for spotting it anyway.
